test/e2e/deploy_filesystem_test.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2023 VMware, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package e2e
+
+import (
+	"context"
+	"fmt"
+	"testing"
+	"testing/fstest"
+	"time"
+
+	"github.com/cppforlife/go-cli-ui/ui"
+	"github.com/google/go-cmp/cmp"
+	"github.com/stretchr/testify/require"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"github.com/vmware-tanzu/carvel-kapp/pkg/kapp/cmd/app"
+	"github.com/vmware-tanzu/carvel-kapp/pkg/kapp/cmd/core"
+	"github.com/vmware-tanzu/carvel-kapp/pkg/kapp/logger"
+)
+
+func TestDeployFilesystem(t *testing.T) {
+	env := BuildEnv(t)
+	testLogger := Logger{}
+	kapp := Kapp{t, env.Namespace, env.KappBinaryPath, testLogger}
+	appName := "test-deploy-filesystem"
+	cleanUp := func() {
+		kapp.Run([]string{"delete", "-a", appName})
+	}
+
+	cleanUp()
+	t.Cleanup(cleanUp)
+
+	theUI := ui.NewConfUI(ui.NewNoopLogger())
+	theUI.EnableNonInteractive()
+
+	configFactory := core.NewConfigFactoryImpl()
+	configFactory.ConfigureClient(100, 200)
+
+	// We don't need to customize any of these, but we'll get panics if they're nil
+	configFactory.ConfigurePathResolver(func() (string, error) {
+		return "", nil
+	})
+	configFactory.ConfigureContextResolver(func() (string, error) {
+		return "", nil
+	})
+	configFactory.ConfigureYAMLResolver(func() (string, error) {
+		return "", nil
+	})
+
+	depsFactory := core.NewDepsFactoryImpl(configFactory, theUI)
+	log := logger.NewUILogger(theUI)
+
+	deployOptions := app.NewDeployOptions(theUI, depsFactory, log)
+	deployOptions.AppFlags.NamespaceFlags.Name = env.Namespace
+	deployOptions.AppFlags.Name = appName
+	deployOptions.FileFlags.Files = []string{
+		"dir1",            // directory
+		"file1.yaml",      // file in root of fs
+		"dir2/file2.yaml", // file in subdir
+		"dir3",            // another directory
+	}
+
+	flagsFactory := core.NewFlagsFactory(configFactory, depsFactory)
+
+	deployCmd := app.NewDeployCmd(deployOptions, flagsFactory)
+
+	now := time.Now().Unix()
```

Comment:
**question**: I was really curious about why a label was being added and then noticed that we are looking for the cms in all the namespaces during validation and hence to avoid any other cms present, using a label makes sense. But then I got curious about why we need to look into all the namespaces, maybe just the env.Namespace should be enough? (Or maybe I am missing some use case that we are trying to cover)

Reply:
I can switch it to just look in env.Namespace